p2p/src/cybershake.rs

```diff
@@ -218,63 +239,134 @@ where
-    pub async fn receive_message(&mut self) -> Result<Vec<u8>, Error> {
-        let mut lenbuf = [0u8; 8];
-        let seq = self.seq;
-        self.seq += 1;
-        self.reader.read_exact(&mut lenbuf[..]).await?;
-        let len = LittleEndian::read_u64(&lenbuf) as usize;
-
-        // length must include IV prefix (16 bytes)
-        if len < 16 {
-            return Err(Error::ProtocolError);
-        }
-        // Check the message length and fail before changing any of the remaining state.
-        if (len - 16) > self.message_maxlen {
-            return Err(Error::MessageTooLong(len - 16));
+impl<W: AsyncWrite + Unpin> AsyncWrite for Outgoing<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let me = self.get_mut();
+
+        ready!(me.flush_pending_ciphertext(cx));
+
+        if me.buf.len() + buf.len() >= PLAINTEXT_BUF_SIZE as usize + 2 {
+            // plaintext_buf has BUF_SIZE size, so subtract with overflow will be never.
+            let size_to_write = PLAINTEXT_BUF_SIZE as usize + 2 - me.buf.len();
+            me.buf.extend_from_slice(&buf[..size_to_write]);
+            me.cipher_buf();
+            ready!(me.flush_pending_ciphertext(cx));
+            Poll::Ready(Ok(size_to_write))
+        } else {
+            me.buf.extend_from_slice(buf);
+            Poll::Ready(Ok(buf.len()))
         }
+    }
 
-        let mut ciphertext = Vec::with_capacity(len);
-        ciphertext.resize(len, 0u8);
-        self.reader.read_exact(&mut ciphertext[..]).await?;
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        let me = self.get_mut();
+        if me.buf.len() == 0 {
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::WriteZero,
+                "attempt to write empty message",
+            )));
+        } else {
+            me.cipher_buf();
+        }
+        me.flush_pending_ciphertext(cx)
+    }
 
-        self.kdf.append_u64(b"seq", seq);
-        let mut key = [0u8; 32];
-        self.kdf.challenge_bytes(b"key", &mut key);
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        let me = self.get_mut();
+        if !me.flushing {
+            me.cipher_buf();
+        }
+        ready!(me.flush_pending_ciphertext(cx));
+        me.writer.as_mut().poll_shutdown(cx)
+    }
+}
 
-        let ad = encode_u64le(seq);
+impl<W: AsyncRead + Unpin> Incoming<W> {
+    pub async fn receive_message(&mut self) -> Result<Vec<u8>, Error> {
+        let mut buf = vec![0u8; 4096];
 
-        let plaintext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
-            .decrypt(&[&ad], &ciphertext)
-            .map_err(|_| Error::ProtocolError)?;
+        let len = self.read(&mut buf).await.map_err(|e| match e.kind() {
+            io::ErrorKind::InvalidData => Error::ProtocolError,
```

Comment:
Why not wrap this kind into our Error::IoError, like all other io::ErrorKind variants?

Reply:
I choose io::ErrorKind::InvalidData as marker which signalize that error has been occured when we deciphered buffer.

Reply:
let's kill our Error and use io::Error with InvalidData kind for version mismatch and encryption failures.

Reply:
removed